server.go

```diff
@@ -2041,37 +2053,53 @@ func (s *Server) serveConn(c net.Conn) (err error) {
 		}
 
 		// 'Expect: 100-continue' request handling.
-		// See http://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html for details.
+		// See https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.3 for details.
 		if ctx.Request.MayContinue() {
-			// Send 'HTTP/1.1 100 Continue' response.
-			if bw == nil {
-				bw = acquireWriter(ctx)
-			}
-			_, err = bw.Write(strResponseContinue)
-			if err != nil {
-				break
-			}
-			err = bw.Flush()
-			if err != nil {
-				break
-			}
-			if s.ReduceMemoryUsage {
-				releaseWriter(s, bw)
-				bw = nil
-			}
 
-			// Read request body.
-			if br == nil {
-				br = acquireReader(ctx)
-			}
-			err = ctx.Request.ContinueReadBody(br, maxRequestBodySize, !s.DisablePreParseMultipartForm)
-			if (s.ReduceMemoryUsage && br.Buffered() == 0) || err != nil {
-				releaseReader(s, br)
-				br = nil
+			//Allow the ability to deny reading the incoming request body
+			if s.DenyRequest != nil {
+				if deniedRequest = s.DenyRequest(&ctx.Request.Header); deniedRequest {
+					if br != nil {
+						br.Reset(ctx.c)
```

Comment:
@erikdubbelboer After debugging the strange relationship between the headers and request bodies, this ended up being the "correct" way to reproduce whatever is currently happening inside of serveConn.  

Someone more familiar with the code definitely needs to weigh in on this piece here.

Reply:
Is this really needed? Shouldn't the reader be empty at this point?

Reply:
From what I have tested it is 100% needed.  If not the loop will come back around, attempt a call on ````if err = ctx.Request.Header.Read(br); err == nil {```` and fail with ````server_test.go:1705: Unexpected error from serveConn: error when reading request headers: EOF.````

I believe this is because br is not nil so
````
br = acquireReader(ctx)
````
Does not get called, which means it's already in a place somewhere but hasn't been fully read or reset.  The function that would normally change it's position or use it later on ````ctx.Request.ContinueReadBody```` hasn't been called in this case because the request was abandoned.

At least that's what I think?